pkg/kubelet/apis/config/types.go

```diff
@@ -661,8 +661,8 @@ type ShutdownGracePeriodByPodPriority struct {
 
 type MemorySwapConfiguration struct {
 	// swapBehavior configures swap memory available to container workloads. May be one of
-	// "", "LimitedSwap": workload combined memory and swap usage cannot exceed pod memory limit
-	// "UnlimitedSwap": workloads can use unlimited swap, up to the allocatable limit.
+	// "", "NoSwap": workloads can not use swap, default option.
+	// "LimitedSwap": workload swap usage is limited. The swap limit is proportionate to the container's memory request.
 	// +featureGate=NodeSwap
 	// +optional
 	SwapBehavior string
```

Comment:
There are two ways people can use the beta swap feature in 1.29 that this change will disrupt:

1. running 1.29 kubelets with `--fail-swap-on=false --feature-gates=NodeSwap=true` and `swapBehavior: UnlimitedSwap`
   * configurations like this will fail to start in 1.30 with validation errors because we've dropped UnlimitedSwap support
   * it sounds like the feedback from the beta was that we couldn't safely offer this option
   * this is at least very clear at startup
2. running 1.29 kubelets with `--fail-swap-on=false --feature-gates=NodeSwap=true` and not specifying `swapBehavior` at all in the kubelet config
  * configurations like this will switch from running LimitedSwap to not enabling swap at all
  * this is likely to be missed by people and seems more problematic to me

I would expect people who didn't opt into swap previously to not accidentally start using it, and people who *did* opt into swap in 1.29 to not accidentally *stop* using it.

Reply:
Is this something we can solve with documentation? I'm not really sure how to handle the second point to be honest.

@mrunalp @SergeyKanzhelev WDYT?

By removing UnlimitedSwap, we would only allow people to use Swap if their pods are burstable. So we expect a reduced scope of users but it does impact those who used UnlimitedSwap.

Reply:
What if we don't default and require the swap mode to be explicitly set? If it isn't set in the 2nd case above then we can error out.

Reply:
We can also document how to verify that swap is available in your cluster. Does it show up in `.status.capacity` for the Node? (if not, I wonder if it should appear somewhere else).

Reply:
I added a PR to nfd to add a swap label to nodes if swap exists. That of course requires NFD.

For kube, we worry about putting that in .status.capacity as only burstable pods could utilize swap and you are not guaranteed to use swap. Its only at pressure.

We did add stats to summary stats in kubelet that one could use to verify swap capacity.

Reply:
>  we worry about putting that in `.status.capacity` as only burstable pods could utilize swap and you are not guaranteed to use swap

Related, I wonder if we should report swap pressure as a condition (we don't necessarily have to trigger a taint when the condition occurs).

Reply:
Yea that is one thing we may consider in future KEP work. We decided to not support UnlimitedSwap because it isn't clear what it means to have swap pressure and what to do in that scenario. And with UnlimtedSwap you can easily use up all the swap.

@iholder101, @fabiand and I talked a bit about this actually. We are still researching what it means to have swap pressure and what kind of evictions we would want to take. Swap is highly dependent on the kernel and the type of swap (disk, memory).

Reply:
As a starting point, how about setting a threshold for page file use and mapping that to a condition but taking no further action? The default can be to set the condition at 100% use, with a configuration setting to allow a lower figure.

At least then, if someone wants nodes tainted when the condition happens, they can make their own controller and supply their own taint. And of course the mechanism to evict based on taints is already well defined.

Reply:
Main thing we find with users of swap, they find that the eviction manager is getting in the way. So a few users have disabled memory evictions if swap is enabled.

https://github.com/kubernetes/enhancements/issues/2400#issuecomment-1943207418

Its not an easier answer at the moment so we took a more conservative approach with this kep and focused on a smaller use case for LimitedSwap.

Reply:
@liggitt WDYT of @mrunalp suggestion?

I drafted up the work https://github.com/kubernetes/kubernetes/pull/123645 in case we decide to go that route. I'll merge that work into here but wanted to see if this is the path we would take.

We would fail the case "" and add a validation error on kubelet startup. We default to NoSwap starting from 1.30.

Reply:
and @sftim We are keeping this feature in beta and we are open to SwapPressure but I think I'd like some more suggestions from users on what they would want.

Is there an objection to adding that as a followup?

We would like to add this functionality as this would block us on GAing eventually. We discovered that CI/users do in fact use `--fail-swap-on=false` and they don't expect swap to be used.

Reply:
> configurations like this will switch from running LimitedSwap to not enabling swap at all

A small note: in 1.29 the default behavior was actually UnlimitedSwap, which is now dropped, therefore we have no other option but to change the default. IMHO doing so for a feature that did not graduate to full Beta yet is reasonable and we should be able to get around it with documentation.

> What if we don't default and require the swap mode to be explicitly set? If it isn't set in the 2nd case above then we can error out.

While it's true that @mrunalp's suggestion would be more visible, it's probably worse for the longer term than what's suggested in this PR since:
1. NodeSwap feature gate would be on by default. So if `--fail-swap-on=false` is provided kubelet would fail. IMO it's somewhat of an odd behavior, since the user explicitly asked kubelet not to fail when swap is on, but now he needs to take two steps to make this happen. We can instead deprecate `--fail-swap-on` and make it a no-op, but I don't really see why this is better than what's proposed by this PR.
2. This would cause (our and possibly others') CI to fail, since swap behavior is not provided.

> Yea that is one thing we may consider in future KEP work.

:+1:
I think that this KEP provides a stable, yet minimal and basic use of swap. There are many further areas to explore and expand (evictions, more customized control, better visibility, and much more), which I think should be done in a follow-up KEP.

This way we would able to provide k8s users with a stable yet basic use of swap while we can gain feedback,  inputs, and understand use-cases better in order to present more advanced swap in the future.

Reply:
Instead of a string we can create a first class type.

```
type SwapBehavior string

const (
  UnlimitedSwap = "UnlimitedSwap"
  NoSwap = "NoSwap"
)

...
```

Reply:
I would love that but I was thinking that would be a breaking change…

we should’ve used a first class type in alpha as that would make support easier.

Reply:
We prefer not to make breaking changes, but we can (the feature is not GA).

Reply:
> What if we don't default and require the swap mode to be explicitly set? If it isn't set in the 2nd case above then we can error out.

I guess requiring someone who has opted into running with swap (by changing the `--fail-swap-on=false` arg) to explicitly indicate the workload swap behavior they want is probably the least bad option here.

There's three user config knobs related to swap today:

1. `featureGates["NodeSwap"]` (or `--feature-gates=NodeSwap=...` kubelet flag)
  * defaults to false <1.30
  * desired to default to true >=1.30 (as ~all feature gates are expected to do eventually)
2. `failSwapOn` (or `--fail-swap-on` kubelet flag)
  * defaults true
3. `memorySwap.swapBehavior`
  * defaults to UnlimitedSwap <1.30
  * desired to not default to UnlimitedSwap >=1.30

If a future release wants to change `--fail-swap-on` to default to `false`, there will be another can of worms here.

> @kannon92:
> We discovered that CI/users do in fact use --fail-swap-on=false and they don't expect swap to be used.

Oof... I wonder how many of them understand implications like #105978

> @iholder101:
> A small note: in 1.29 the default behavior was actually UnlimitedSwap, which is now dropped, therefore we have no other option but to change the default. IMHO doing so for a feature that did not graduate to full Beta yet is reasonable and we should be able to get around it with documentation.

Good to know.

> @iholder101:
> We can instead deprecate --fail-swap-on and make it a no-op, but I don't really see why this is better than what's proposed by this PR.

Changing the default for `--fail-swap-on` or making it a no-op is a different discussion. I expected a safer resolution of #105978 before defaulting the kubelet to be willing to create tmpfs volumes that could actually be persisted to disk.

Reply:
Yea, we discovered this problem with trying to fail this feature for cgroupv1. We figured we could use --failSwapOn=false as who is using that field if we don't have swap enabled. Well kind jobs in our CI provision failSwap=false and are using cgroup v1. So that was fun.

We figured that defaulting NoSwap can at least not break the failSwapOn=false usecase with cgroupv1. I sorta view failSwap as separate to this feature at the moment. It was thought in the early stages of this feature that we can safely view it as a toggle for this feature but we discovered problems.

Its a prereq for this feature but it existed before this feature so we can't view it as a toggle for this feature.

Reply:
> this is likely to be missed by people and seems more problematic to me

@liggitt I do not disagree, but wasn't one goal of this beta2 to create more certainty and safety for users?
To me the decision about the new defaults (NoSwap) and removing features (UnlimitedSwap) seems to be a direct consequence of this research. Aka the goal of beta2 was reached :)

How about adding a warning or info to the logs of the relevant component in order to make this change easier to discover (thus a warning) and to notify the user about these two changes and their mitigation.
After all the mitigation is at least rather easy.

Reply:
I followed up with @liggitt on slack.

> can we detect when they opted into swap with --fail-swap-on=false and no swapBehavior config option was provided, and output a warning at least in the kubelet log, so we aren't silently switching from making workloads use swap to not use swap? 

I am adding a log to make it clear that we are using NoSwap if swapBehavior is not specified. I can't really use --fail-swap-on=false but if fail-swap-on=true is set and we are using swap then kubelet would fail to start so I think that is fine.

Reply:
Added:

`	switch m.memorySwapBehavior {
	default:
		klog.InfoS("NoSwap is set due to memorySwapBehavior not specified", "memorySwapBehavior", m.memorySwapBehavior)
		swapConfigurationHelper.ConfigureNoSwap(lcr)
	}`